src/lib/mongoBackend/mongoUpdateContext.cpp

```diff
@@ -38,12 +38,60 @@
 #include "ngsi/NotifyCondition.h"
 #include "rest/HttpStatusCode.h"
 
+#include "ngsiNotify/QueueNotifier.h"
+
 #include "mongoBackend/MongoGlobal.h"
 #include "mongoBackend/MongoCommonUpdate.h"
 #include "mongoBackend/mongoUpdateContext.h"
 
 
 
+/* ****************************************************************************
+*
+* flowControlAwait -
+*/
+static void flowControlAwait(unsigned int q0, unsigned int notifSent)
+{
+  unsigned int pass = 0;
+  unsigned int accumulatedDelay = 0;
+
+  // target may range from q0 (for fcGauge = 1) to q0 + notifSent (for fcGauge = 0)
+  unsigned int target = q0 + (1 - fcGauge) * notifSent;
+  LM_T(LmtNotifier, ("flow control target is %d", target));
+
+#ifdef UNIT_TEST
+  // not used in test (and the QueueNotifier class cause build problems)
+  unsigned int qi = 0;
+#else
+  // get current size of the queue
+  unsigned int qi = ((QueueNotifier*) getNotifier())->queueSize();
+#endif
+  LM_T(LmtNotifier, ("flow control pass %d, delay %d, notification queue size is: %d",
+                     pass, accumulatedDelay, qi));
+
+
+  while (qi > target)
+  {
+    pass++;
```

Comment:
We should check that the connection with the client is still active or break in negative case. Not sure how to do this... we have a ciP->connection object (a MDH_Connection) which hopefully could be used to check this. However libmicrohttpd library documentation is not very clear on this. Maybe the MHD_get_connection_info() function can be used for this.

Reply:
It is not clear how to implement this at the presesnt moment (suggestions/hints welcome ;) so probably I'll move the text to a code comment with a FIXME mark. It is not a big issue, after all (only involves some inefficiency in the case of client cutting the connection, which is not usual).

Reply:
Comment added in commit 612b80d12a88236b1927641547d83491d95fb3ba